docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -0,0 +1,378 @@
+  return store.Has(types.KeyTrace(denomTraceHash))
+}
+
+// SetTrace sets a new {trace hash -> trace} pair to the store.
+func (k Keeper) SetTrace(ctx Context, denomTrace DenomTrace) {
+  store := ctx.KVStore(k.storeKey)
+  bz := k.cdc.MustMarshalBinaryBare(&denomTrace)
+  store.Set(types.KeyTrace(denomTrace.Hash()), bz)
+}
+```
+
+The problem with this approach is that when a token is received for the first time, the full trace
+info will need to be passed in order to construct the hash and set it to the mapping on the store.
+To mitigate this a new `Trace` field needs to be added to the `FungibleTokenPacketData`:
+
+```protobuf
+message FungibleTokenPacketData {
+  // the tokens to be transferred
+  repeated cosmos.Coin amount = 1 [
+    (gogoproto.nullable)     = false,
+    (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins"
+  ];
+  // coins denomination trace for tracking the source
+  repeated DenomTrace denom_traces = 2;
+  // the sender address
+  string sender = 3;
+  // the recipient address on the destination chain
+  string receiver = 4;
+}
+```
+
+The `MsgTransfer` will validate that the Coins from the `Amount` field contain the hash that is valid:
+
+```golang
+func (msg MsgTransfer) ValidateBasic() error {
+  // ...
+  if len(msg.Amount) != len(msg.DenomTraces) {
+    //  error
+  }
+  for i := range msg.Amount {
+    hash, err := GetTraceHashFromDenom(msg.Amount[i].Denom)
+    if err != nil {
+      return err
+    }
+    if hash != msg.DenomTraces[i].Hash() {
+      // error
+    }
+  }
+}
+```
+
+```golang
+func GetTraceHashFromDenom(rawDenom string) (tmbytes.HexBytes, error) {
+  denomSplit := strings.SplitN(denom, "/", 3)
+
+  switch {
+    case len(denomSplit) == 0:
+      err = Wrap(ErrInvalidDenomForTransfer, denom)
+    case denomSplit[0] == denom:
+      err = Wrapf(ErrInvalidDenomForTransfer, "denomination should be prefixed with the format 'ibc/{hash(trace + \"/\" + %s)}'", denom)
+    case denomSplit[0] != "ibc":
+      err = Wrapf(ErrInvalidDenomForTransfer, "denomination %s must start with 'ibc'", denom)
+    case len(denomSplit) == 2  && len(denomSplit[1]) != 32:
+      err = Wrapf(ErrInvalidDenomForTransfer, "invalid SHA256 hash %s", denomSplit[1])
+    default:
+      err = Wrap(ErrInvalidDenomForTransfer, denom)
+  }
+
+  if err != nil {
+    return nil, err
+  }
+
+  return denomSplit[1], nil
+}
+```
+
+When a fungible token is sent to a sink chain, the trace information needs to be updated with the
+new port and channel identifiers:
+
+The denomination also needs to be updated when token is received on the source chain. This is done during the `SendTransfer`'s `createOutgoingPacket` call and the `OnRecvPacket` call. Below is the updated function that sets the :
+
+```golang
+// createOutgoingPacket
+// ...
+prefix := types.GetDenomPrefix(destinationPort, destinationChannel)
+source := strings.HasPrefix(amount[0].Denom, prefix)
+
+for _, denomTrace := range denomTraces {
```

Comment:
Are you getting these denomTraces from the packet?

If so, shouldn't there only be 1 since only 1 coin denom/packet is supported now. Or is this ADR intended to work for future multi denom packets?

Reply:
only one denom trace per packet (needs to match the token len). Will update accordingly